lib/chef/provider/aws_ebs_volume.rb

```diff
@@ -27,14 +37,71 @@ class Chef::Provider::AwsEbsVolume < Chef::Provider::AwsProvider
 
   action :delete do
     if existing_volume
-      converge_by "Deleting EBS volume #{fqn} in #{new_resource.region_name}" do
-        existing_volume.delete
+      begin
+        converge_by "Deleting EBS volume #{fqn} in #{new_resource.region_name}" do
+          existing_volume.delete
+
+          Retryable.retryable(:tries => 30, :sleep => 2, :on => TimeoutError) do
+            raise TimeoutError,
+              "Timed out waiting for volume #{new_resource.volume_id} to be deleted." if existing_volume.exists?
+          end
+        end
       end
     end
 
     new_resource.delete
   end
 
+  action :attach do
+    if existing_volume.nil?
+      raise RuntimeError, "Volume #{new_resource.name} does not exist. Can not attempt attachment to an instance."
+    elsif volume_attached_to_configured_instance?
+      Chef::Log.debug "Volume already attached to expected instance."
+    # if volume is attached, but to a different instance allow aws sdk to handle the error
+    else
+      converge_by "Attaching EBS volume #{fqn} in #{new_resource.region_name} to instance #{new_resource.instance_id}" do
+
+        existing_volume.attach_to(
+          ec2.instances[new_resource.instance_id],
+          new_resource.device
+        )
+        new_resource.attached_to_instance new_resource.instance_id
+        new_resource.attached_to_device new_resource.device
+
+        wait_for_volume_status(:in_use)
```

Comment:
We are not checking for the right thing here: we want to wait for the volume to be attached to this particular instance, yes?  What happens if you call attach and then someone else calls detach/attach before you check for in use?